### PR TITLE
Add tracing aware executor utils

### PR DESCRIPTION
**Pull request by markelliot**

### CODE REVIEW

**Comment by markelliot** [comment]
@uschi2000

**Comment by markelliot** [comment]
some atlasdb prior art from its original code base:
https://github.com/palantir/atlasdb/blob/develop/commons-executors/src/main/java/com/palantir/common/concurrent/ExecutorInheritableThreadLocal.java
https://github.com/palantir/atlasdb/blob/develop/commons-executors/src/main/java/com/palantir/common/concurrent/PTExecutors.java
(these would be the more generic approach)

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareRunnable.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting1.tracing;
+
+import java.util.Deque;
+
+public final class TracingAwareRunnable implements Runnable {
```
Comment (uschi2000) [inline tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareRunnable.java:21]: can this be package-private?

File: tracing/src/main/java/com/palantir/remoting1/tracing/Traces.java [hunk 2]
```diff
@@ -47,6 +49,21 @@
...
+     */
+    static Deque<TraceState> getCopyOfState() {
+        return new ArrayDeque<>(STATE.get());
+    }
+
+    /**
+     * Package-local state forcing mechanism for thread inheritance.
+     */
+    static void forceState(Deque<TraceState> stateToForce) {
+        STATE.get().remove();
```
Comment (uschi2000) [inline tracing/src/main/java/com/palantir/remoting1/tracing/Traces.java:63]: this just removes one TraceState, right? don't you need to remove all of them?

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.remoting1.tracing;
+
+import java.util.Deque;
+import java.util.concurrent.Callable;
+
+public final class TracingAwareCallable<V> implements Callable<V> {
```
Comment (uschi2000) [inline tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java:22]: can be package-private?

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final Callable<V> delegate;
+    private final Deque<TraceState> state;
+
+    public TracingAwareCallable(Callable<V> delegate) {
+        this.delegate = delegate;
+        this.state = Traces.getCopyOfState();
```
Comment (uschi2000) [inline tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java:29]: I think it'll be more clear if you inject the `state`

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareExecutorService.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    @Override
+    public final Future<?> submit(Runnable task) {
+        return delegate.submit(new TracingAwareRunnable(task));
+    }
+
+    @Override
+    public final <T> List<Future<T>> invokeAll(Collection<? extends Callable<T>> tasks) throws InterruptedException {
+        return delegate.invokeAll(Collections2.transform(tasks, new TracedCallableConverter<T>()));
```
Comment (uschi2000) [inline tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareExecutorService.java:80]: use static instance of converter

Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareExecutorService.java:80]: oh, actually that may not work that nicely due to generics?

Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareExecutorService.java:80]: this code path is going to be quite hot, so maybe it's worthing avoiding the object creation

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java (hunk 4 above)
Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java:29]: alternatively, leave a class-level comment: 
```A {@link Callable} that uses the thread-local tracing state at the time of it's construction during its {@link #call execution}. The {@link TracingAwareCallable} constructor is called by a tracing-aware executor service on the same thread on which a user creates a Callable; the {@link call} function is executed on an arbitrary (likely different) thread with different tracing state. In order to execute the task with the original (and intuitively expected) tracing state, we remember the original state and set it for the duration of the {@link #call execution}.

```
```

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareRunnable.java (hunk 1 above)
Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareRunnable.java:21]: add a `See {@link TracingAwareCallable}` comment

File: tracing/src/main/java/com/palantir/remoting1/tracing/Traces.java (hunk 2 above)
Comment (uschi2000) [reply tracing/src/main/java/com/palantir/remoting1/tracing/Traces.java:63]: we should some tests for this ...

**Review by uschi2000 — CHANGES_REQUESTED** [review]

**Comment by markelliot** [comment]
do you want to take this over?

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java (hunk 3 above)
Comment (markelliot) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java:22]: Probably want to provide a way to wrap a `Callable<V>` generally, fine with that being a utility class that hides this implementation. Maybe the utility class should nest these implementations as private static classes?

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java (hunk 4 above)
Comment (markelliot) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java:29]: utility class mentioned in previous comment would make it more apparent how to do the wrapping and more obvious that the state is being injected. The main thing is that a caller that needs to wrap their `Callable<V>` submission should be able to do so without having to know about how to grab present state.

Comment (markelliot) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareCallable.java:29]: Speaking of which, I suspect we probably want to do something in the ExecutorService implementations that checks if the callable being passed is already an implementation of this, so we can avoid the duplicate work.

File: tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareExecutorService.java (hunk 5 above)
Comment (markelliot) [reply tracing/src/main/java/com/palantir/remoting1/tracing/TracingAwareExecutorService.java:80]: unfortunately the `<T>` makes this very hard. `invokeAll` likely isn't a hot code path, though, the `Callable<T>`s are likely the bulk of the work.

**Comment by markelliot** [comment]
obe

**Pull request closed without merging**